src/vision/deliverables/async_fn/async_drop.md

```diff
@@ -0,0 +1,115 @@
+
+Making this work requires [inline async fn]. This is because Rust presently assumes *all* types are croppable. Consider a function `foo`:
+
+```rust
+async fn foo<T>(x: T) {}
+```
+
+Here, we will drop `x` when `foo` returns, but we do not know whether `T` implements `AsyncDrop` or not, and we won't know until monomorphization. However, to know whether the resulting future for `foo(x)` is `Send`, we have to know whether the code that drops `x` will be send. Using an inline function, we know that `T: Send` implies that the async drop future for `T` is `Send`.
+
+Another argument in favor of [inline async fn] is that dropping ought not to create a lot more memory.
+
+[inline async fn]: ./inline_async_fn.md
+
+### Explicit async drop
+
+We should have a `std::mem::async_drop` analogous to `std::mem::drop`:
+
+```rust
+async fn async_drop<T>(x: T) { }
+```
+
+### Implicit await points
+
+When you run async drop glue, there is an implicit await point. Consider this example:
+
+```rust
+async fn foo(dbc: DatabaseConnection) -> io::Result<()> {
+    let data = socket().read().await?;
+    dbc.write(data).await?;
+}
+```
+
+Here, presuming that `DatabaseConnection` implements `AsyncDrop`, there are actually a number of async drops occurring:
+
+```rust
+async fn foo(dbc: DatabaseConnection) -> io::Result<()> {
+    let data = match socket().read().await {
+        Ok(v) => v,
+        Err(e) => {
+            std::mem::async_drop(dbc).await;
+            return e;
+        }
+    };
+    let () = match dbc.write(data).await? {
+        Ok(()) => (),
+        Err(e) => {
+            std::mem::async_drop(dbc).await;
+            return e;
+        }
+    };
+    std::mem::async_drop(dbc).await;
+}
+```
+
+As this example shows, there are important ergonomic benefits here to implicit async drop, and it also ensures that async and sync code work in analogous ways. However, implicit await points can be a hazard for some applications, where it is important to identify all await points explicitly (for example, authors of embedded applications use await points to reason about what values will be stored in the resulting future vs the stack of the poll function). To further complicate things, async-drop doesn't only execute at the end of a block or an "abrupt" expression like `?`: async-drop can also execute at the end of every statement, given temporary values.
+
+The best solution here is unclear. We could have an "allow-by-default" lint encouraging explicit use of `async_drop`, but as the code above shows, the result may be highly unergonomic (also, imagine how it looks as the number of variables requiring async-drop grows).
+
+Another option is to target the problem from another angle, for example by adding lints to identify when large values are stored in a future or on the stack, or to allow developers to tag local variables that they expect to be stored on the stack, and have the compiler warn them if this turns out to not be true.
```

Comment:
Does the second idea imply some kind of implicit allocation?

Reply:
Nothing in this paragraph implies implicit allocation to me -- just warnings when things didn't turn out how you might expect. You can choose how to resolve that.

Reply:
Added a sentence that hopefully clarifies.